crates/forge-runner/src/test_case_summary.rs

```diff
@@ -264,23 +268,78 @@ impl TestCaseSummary<Single> {
     }
 }
 
-/// Helper function to build `readable_text` from a run data.
+/// Helper function to build readable_text from a run data.
 fn build_readable_text(data: &Vec<Felt252>) -> Option<String> {
-    let mut readable_text = String::new();
+    let mut data_iter = data.clone().into_iter().peekable();
+    let mut items = Vec::new();
 
-    for felt in data {
-        readable_text.push_str(&format!("\n    original value: [{felt}]"));
-        if let Some(short_string) = as_cairo_short_string(felt) {
-            readable_text.push_str(&format!(", converted to a string: [{short_string}]"));
+    while let Some(item) = format_next_felt(&mut data_iter) {
+        items.push(format!("\n{item}"));
+    }
+
+    match items[..] {
+        [] => None,
+        _ => {
+            let mut result = items.join("").to_string().replace('\n', "\n    ");
+            result.push('\n');
+            Some(result)
+        }
+    }
+}
+
+fn format_next_felt<T>(values: &mut T) -> Option<String>
+where
+    T: Iterator<Item = Felt252> + Clone,
+{
+    let first_felt = values.next()?;
+
+    if first_felt == felt_str!(BYTE_ARRAY_MAGIC, 16) {
+        if let Some(string) = try_format_string(values) {
+            return Some(string);
         }
     }
 
-    if readable_text.is_empty() {
-        None
-    } else {
-        readable_text.push('\n');
-        Some(readable_text)
+    if let Some(short_string) = as_cairo_short_string(&first_felt) {
+        return Some(format!(
+            "original value: [{first_felt}], converted to a string: [{short_string}]"
+        ));
     }
+    Some(format!("original value: [{first_felt}]"))
+}
+
+// https://github.com/starkware-libs/cairo/blob/1a4ccb01f6fb0a74bfc23238ca1088fe2659d087/crates/cairo-lang-runner/src/casm_run/mod.rs#L2227
+
+/// Tries to format a string, represented as a sequence of `Felt252`s.
+/// If the sequence is not a valid serialization of a ByteArray, returns None and doesn't change the
+/// given iterator (`values`).
+fn try_format_string<T>(values: &mut T) -> Option<String>
```

Comment:
Unfortunately it's a private function

Reply:
Can we have an issue to make it public in the compiler?